### PR TITLE
[SPARK-28040][SPARK-28070][R] Write type object s3

### PR DESCRIPTION
**Pull request by MichaelChirico**
Re-filing stale https://github.com/apache/spark/pull/24888

### What changes were proposed in this pull request?

Migrate `write*` internals to use an S3 generic

### Why are the changes needed?

Robustness -- old version was missing some cases like `glue` that would be treated properly as `character` by S3 dispatch

### CODE REVIEW

**Comment by MichaelChirico** [comment]
@HyukjinKwon oh! actually the issue is not transient.

> The repository 'https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/ Release

that file doesn't exist anymore because the `35` folder is no longer the "head" of R -- it's migrated to 40:

https://cloud.r-project.org/bin/linux/ubuntu/bionic-cran40/

**Comment by HyukjinKwon** [comment]
Thanks for pointing it out. I just opened a PR :-)

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -105,7 +105,7 @@ jobs:
         java-version: '11'
     - name: install R
       run: |
-        echo 'deb https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/' | sudo tee -a /etc/apt/sources.list
+        echo 'deb https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/' | sudo tee -a /etc/apt/sources.list
```
Comment (HyukjinKwon) [inline .github/workflows/master.yml:108]: Let's ignore this for now. I will track it in SPARK-31587

Comment (MichaelChirico) [reply .github/workflows/master.yml:108]: I pushed before I saw your note. Will leave this until that issues fixed & we can update all the other PRs

**Comment by HyukjinKwon** [comment]
cc also @felixcheung and @shivaram

**Comment by MichaelChirico** [comment]
Latest commit was vetted for regression by comparing output of a bunch of common / edge case objects; those tests are here, might make sense to include them as unit tests:

https://gist.github.com/MichaelChirico/58327b30aa865fd8bb234b7b98fa6df9

**Comment by MichaelChirico** [comment]
alright i'm stuck. what could cause this?

> R worker produced errors: 
  no applicable method for 'writeObject' applied to an object of class "c('integer', 'numeric')"
> Calls: <Anonymous>

**Comment by MichaelChirico** [comment]
> what could cause this?

I believe it was the use in `worker.R` of `SparkR:::writeObject`; since there were no S3 methods registered in `NAMESPACE` we get that error

**Comment by MichaelChirico** [comment]
@HyukjinKwon 7fccb3ab4b9658aa57a1ebab3aedc0c1633d5bc8 failed linting on Jenkins, but not on GH Actions. The error:

```
inst/worker/worker.R:81:62: style: Opening curly braces should never go on their own line and should always be followed by a new line.
    arrow = SparkR:::writeArrowSerialize(output, outputCon), { # brace here for linter...
                                                             ^
lintr checks failed.
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/dev/lint-r ; received return code 1
```

Two things:

 1. Is the `lintr` version on Jenkins up to date? I see there was something added to `lintr` in Sep '19 about this, so if the Jenkins version is `<= 1.0.3`, that could be the source
 2. Is it necessary to have both Jenkins and GitHub running the linting check?

**Comment by MichaelChirico** [comment]
Alright, completely and totally stuck now.

Could go for an extra pair of eyes / advice on how to further debug @HyukjinKwon .

Have done extensive regression testing, see https://gist.github.com/MichaelChirico/58327b30aa865fd8bb234b7b98fa6df9

Two types of test (1) test a bunch of common objects to be sure `writeObject` is producing the same bits on `master` and this branch (2) ran `pkg/run-tests.sh` where I added some code to `writeObject` to `save()` the input of _all_ instances of `writeObject` encountered during the tests, then do the old-v-new comparison on all of those.

Both are successful, so as near as I can tell, the new and old `writeObject` are always producing the same output.

But still we're failing (it seems a timeout error??)

I'm also not able to reproduce the same errors as Jenkins or Appveyor (which have different errors) locally, so it's hard to know how to proceed.

I'm also not able to run `pkg/run-tests.sh` on `master` locally either, FWIW (Spark itself builds under `./build/mvn -DskipTests clean package`), further complicating the dev process...

**Comment by MichaelChirico** [comment]
For the Jenkins `lintr`, I see now it's 2.0.0 installed from GitHub; this can probably be handled by CRAN now right? 2.0.1 is on CRAN

**Comment by MichaelChirico** [comment]
New & old `writeObject` differ for this input:

```
x = c(1:2, NA)

# old
rc = rawConnection(raw(), 'rb')
writeObject(rc, x)
rawConnectionValue(rc)
# 6169000000030000000100000002
close(rc)

# new
rc = rawConnection(raw(), 'rb')
writeObject(x, rc)
rawConnectionValue(rc)
# 616900000003000000010000000280000000
close(rc)
```

However, the old code throws a warning in this case, so I doubt that such input was observed previously.

**Comment by felixcheung** [comment]
I didn’t review but I recall something similar this to support glue like 2 or 3 years ago? This serde is very sensitive - we probably need (to add) a lot more tests on this (existing tests are not very comprehensive)

**Comment by MichaelChirico** [comment]
Thanks @felixcheung indeed that was me filing that issue about `glue` 😅

This PR is intended to be a much more general solution to that issue -- a `glue` object would be passed to `writeObject.character` under this PR.

There are already a bunch of tests you could basically drop in from the gist above.

My suspicion at the moment is that the usage in `worker.R` is the issue, but I'm having trouble debugging that. Any suggestion for debugging the executor process?

**Comment by danzafar** [comment]
why not just coerce inputs to class `character`?

```
SparkR:::sql.default <- function (sqlQuery) {
    sparkSession <- getSparkSession()
    sdf <- callJMethod(sparkSession, "sql", as.character(sqlQuery))
    dataFrame(sdf)
}
```

**Pull request closed without merging**